### PR TITLE
feat: add Yarn and pnpm cache providers

### PR DESCRIPTION
**Pull request by juliensimon**
## Summary

- Add **Yarn provider** supporting both Classic (`~/Library/Caches/Yarn/v6`) and Berry (`.yarn/berry/cache`) cache formats with semantic naming, package ID extraction, and scoped package resolution
- Add **pnpm provider** supporting content-addressable stores (v3–v10+) and virtual stores (`node_modules/.pnpm`), with index file parsing for vulnerability/outdated scanning
- **Auto-detect** Yarn and pnpm cache paths at startup via CLI probes (`yarn cache dir`, `pnpm store path`) with platform-specific fallbacks
- Fix **tree panel alignment** when Unicode status icons (⚠↓) are present
- Deduplicate overlapping cache roots (e.g. `~/Library/Caches/Yarn` vs `~/Library/Caches/Yarn/v6`)

## What's included

- `src/providers/yarn.rs` — Yarn Classic + Berry filename parsing, `semantic_name`, `package_id`, `metadata`
- `src/providers/pnpm.rs` — virtual store + v10 index file parsing, content-addressable store display
- `src/config.rs` — CLI-based cache path detection, `is_ancestor_or_descendant` dedup helper
- `src/providers/mod.rs` — `CacheKind::Yarn`/`CacheKind::Pnpm` detection, dispatch, safety levels
- `src/ui/tree_panel.rs` — Unicode icon width fix (`chars().count()` vs byte `len()`)
- `tests/integration.rs` — fixture-based integration tests for both providers
- `tests/e2e_js_providers.rs` — E2E tests using real `yarn`/`pnpm` installs
- `.github/workflows/ci.yml` — new E2E test job for JS provider tests

## Test plan

- [x] 1128 unit/integration tests passing
- [x] Clippy clean, cargo fmt clean
- [x] Adversarial tests for edge cases (injection, malformed filenames, empty inputs)
- [x] E2E tests install real packages and verify provider output
- [x] Manual testing: pnpm store shows vulnerability scanning (lodash CVEs detected)
- [x] Manual testing: tree alignment correct with/without status icons
- [ ] CI passes on push

Closes #1

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/ui/tree_panel.rs [hunk 1]
```diff
@@ -72,7 +72,9 @@ pub fn render(
         let name = &node.name;
 
         // Calculate available space for name
-        let prefix_len = indent.len() + arrow.len() + marker.len() + status_icon.len();
+        // Use display width for Unicode icons (⚠ and ↓ are each 1 column but multi-byte)
+        let icon_display_width = status_icon.chars().count();
+        let prefix_len = indent.len() + arrow.len() + marker.len() + icon_display_width;
         let size_len = size_str.len() + 1; // +1 for padding
         let available = width.saturating_sub(prefix_len + size_len + 1);
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/ui/tree_panel.rs:82]: Fixed in b66433e — now using `chars().count()` consistently for all prefix segments (`arrow`, `marker`, and `status_icon`), not just the status icon.

File: docs/superpowers/specs/2026-04-08-yarn-pnpm-providers-design.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+### Metadata
+
+- Package format (tgz vs zip)
+- Yarn version detected (Classic vs Berry)
+- File size
+- Install script detection (if extractable)
+
+### Safety
+
+All Yarn cache entries: `SafetyLevel::Safe` (re-downloadable from npm registry).
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply docs/superpowers/specs/2026-04-08-yarn-pnpm-providers-design.md]: Good catch. Fixed locally in the design doc (gitignored so not pushed) — Berry project-local caches under `.yarn/cache/` are correctly `SafetyLevel::Caution` in the implementation.

File: docs/superpowers/specs/2026-04-08-yarn-pnpm-providers-design.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+## Auto-Detection of Cache Paths (`src/config.rs`)
+
+### CLI Probing (at startup)
+
+If tool is on PATH:
+- `yarn cache dir` -> Yarn 1 global cache path
+- `yarn config get cacheFolder` -> Yarn 2+ cache path
+- `pnpm store path` -> pnpm global store path
+
+Add discovered paths to scan roots. Timeout: 2 seconds per command. Failures are silent (tool may not be installed).
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply docs/superpowers/specs/2026-04-08-yarn-pnpm-providers-design.md]: Correct — there's no timeout implemented. Fixed locally in the design doc (gitignored) to remove the false claim.

File: src/providers/mod.rs [hunk 4]
```diff
@@ -168,17 +208,34 @@ pub fn upgrade_command(kind: CacheKind, name: &str, version: &str) -> Option<Str
...
+        CacheKind::Pnpm => {
+            if path.to_string_lossy().contains("node_modules/.pnpm") {
+                SafetyLevel::Caution
+            } else {
+                SafetyLevel::Safe
+            }
+        }
+        CacheKind::Yarn => {
+            // Berry project-local caches (.yarn/cache/) may be committed to git (zero-install)
+            if path.to_string_lossy().contains(".yarn/cache") {
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/providers/mod.rs]: Fair point — `detect()` already handles both separators. Fixed in b66433e to check both `.yarn/cache` and `.yarn\\cache` in `safety()` for consistency.

File: tests/integration.rs [hunk 5]
```diff
@@ -718,3 +763,183 @@ fn npm_dep_depth_in_metadata() {
...
+        packages
+            .iter()
+            .filter(|(_, id)| id.ecosystem == "npm")
+            .count(),
+        2
+    );
+}
+
+#[test]
+fn yarn_discover_packages_finds_classic_tgz() {
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply tests/integration.rs]: Agreed — renamed to `yarn_discover_packages_finds_classic_cache` in b66433e.

File: src/providers/yarn.rs [hunk 6]
```diff
@@ -0,0 +1,1131 @@
...
+        || path_str.contains(".cache/yarn")
+        || path_str.contains("Library/Caches/Yarn")
+        || path_str.contains(".yarn/cache")
+        || path_str.contains("yarn/berry/cache")
+}
+
+/// Returns true if this is a Yarn Berry (v2+) cache path.
+pub fn is_berry(path: &Path) -> bool {
+    let path_str = path.to_string_lossy();
+    path_str.contains(".yarn/cache") || path_str.contains("berry/cache")
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/providers/yarn.rs:18]: This is a macOS/Linux-only project — Windows is not supported. Added explicit platform statement to README in 56f6894. Not worth the complexity of component-based path matching for a platform we don't target.

File: src/providers/pnpm.rs [hunk 7]
```diff
@@ -0,0 +1,788 @@
...
+pub fn is_pnpm_cache(path: &Path) -> bool {
+    let path_str = path.to_string_lossy();
+    path_str.contains(".pnpm-store")
+        || path_str.contains("pnpm/store")
+        || path_str.contains("node_modules/.pnpm")
+}
+
+/// Returns true if the path is within the pnpm virtual store (node_modules/.pnpm).
+pub fn is_pnpm_virtual_store(path: &Path) -> bool {
+    path.to_string_lossy().contains("node_modules/.pnpm")
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/providers/pnpm.rs:15]: macOS/Linux only — Windows not supported. Documented explicitly in README (56f6894).

File: src/providers/pnpm.rs [hunk 8]
```diff
@@ -0,0 +1,788 @@
...
+            return Some("Package Index".to_string());
+        }
+    }
+
+    // Index files: {hash}-name@version.json → "name version"
+    if name.ends_with(".json") && path_str.contains("/index/") {
+        if let Some(id) = parse_index_filename(&name) {
+            return Some(format!("{} {}", id.name, id.version));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/providers/pnpm.rs:115]: macOS/Linux only — Windows not supported. Documented explicitly in README (56f6894).

File: src/providers/mod.rs [hunk 9]
```diff
@@ -168,17 +208,35 @@ pub fn upgrade_command(kind: CacheKind, name: &str, version: &str) -> Option<Str
...
 /// Get safety level for deletion.
-pub fn safety(kind: CacheKind, _path: &Path) -> SafetyLevel {
+pub fn safety(kind: CacheKind, path: &Path) -> SafetyLevel {
     match kind {
+        CacheKind::Pnpm => {
+            if path.to_string_lossy().contains("node_modules/.pnpm") {
+                SafetyLevel::Caution
+            } else {
+                SafetyLevel::Safe
+            }
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply src/providers/mod.rs:229]: macOS/Linux only — Windows not supported. Documented explicitly in README (56f6894).

File: tests/e2e_js_providers.rs [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+
+    let is_even = packages.iter().find(|(n, _, _)| n == "is-even").unwrap();
+    assert_eq!(is_even.1, "1.0.0");
+
+    let base64 = packages.iter().find(|(n, _, _)| n == "base64-js").unwrap();
+    assert_eq!(base64.1, "1.5.1");
+
+    // Clean up global cache
+    let _ = Command::new("yarn").args(["cache", "clean"]).output();
+}
```
[the remark this answers is not in the document]
Comment (juliensimon) [reply tests/e2e_js_providers.rs]: Good catch. Fixed in 56f6894 — the test now uses `YARN_CACHE_FOLDER` to isolate the cache to a temp directory, and the destructive `yarn cache clean` call has been removed.
